frame/election-provider-multi-phase/src/signed.rs

```diff
@@ -0,0 +1,692 @@
+					break;
+				}
+				Err(_) => {
+					// we assume a worse case feasibility check happened anyhow.
+					weight = weight.saturating_add(feasibility_weight);
+					Self::finalize_signed_phase_reject_solution(&who, deposit);
+					weight = weight
+						.saturating_add(T::WeightInfo::finalize_signed_phase_reject_solution());
+				}
+			}
+		}
+
+		// Any unprocessed solution is not pointless to even ponder upon. Feasible or malicious,
+		// they didn't end up being used. Unreserve the bonds.
+		all_submission.into_iter().for_each(|not_processed| {
+			let SignedSubmission { who, deposit, .. } = not_processed;
+			let _remaining = T::Currency::unreserve(&who, deposit);
+			weight = weight.saturating_add(T::DbWeight::get().writes(1));
+			debug_assert!(_remaining.is_zero());
+		});
+
+		(found_solution, weight)
+	}
+
+	/// Helper function for the case where a solution is accepted in the signed phase.
+	///
+	/// Extracted to facilitate with weight calculation.
+	///
+	/// Infallible
+	pub fn finalize_signed_phase_accept_solution(
+		ready_solution: ReadySolution<T::AccountId>,
+		who: &T::AccountId,
+		deposit: BalanceOf<T>,
+		reward: BalanceOf<T>,
+	) {
+		// write this ready solution.
+		<QueuedSolution<T>>::put(ready_solution);
+
+		// unreserve deposit.
+		let _remaining = T::Currency::unreserve(who, deposit);
+		debug_assert!(_remaining.is_zero());
+
+		// Reward.
+		let positive_imbalance = T::Currency::deposit_creating(who, reward);
+		T::RewardHandler::on_unbalanced(positive_imbalance);
+	}
+
+	/// Helper function for the case where a solution is accepted in the rejected phase.
+	///
+	/// Extracted to facilitate with weight calculation.
+	///
+	/// Infallible
+	pub fn finalize_signed_phase_reject_solution(who: &T::AccountId, deposit: BalanceOf<T>) {
+		let (negative_imbalance, _remaining) = T::Currency::slash_reserved(who, deposit);
+		debug_assert!(_remaining.is_zero());
+		T::SlashHandler::on_unbalanced(negative_imbalance);
+	}
+
+	/// Find a proper position in the queue for the signed queue, whilst maintaining the order of
+	/// solution quality. If insertion was successful, `Some(index)` is returned where index is the
+	/// index of the newly inserted item.
+	///
+	/// Invariant: The returned index is always a valid index in `queue` and can safely be used to
+	/// inspect the newly inserted element.
+	pub fn insert_submission(
+		who: &T::AccountId,
+		queue: &mut Vec<SignedSubmission<T::AccountId, BalanceOf<T>, CompactOf<T>>>,
+		solution: RawSolution<CompactOf<T>>,
+		size: SolutionOrSnapshotSize,
+	) -> Option<usize> {
+		// from the last score, compare and see if the current one is better. If none, then the
+		// awarded index is 0.
+		let outcome = queue
+			.iter()
+			.enumerate()
+			.rev()
+			.find_map(|(i, s)| {
+				if is_score_better::<Perbill>(
+					solution.score,
+					s.solution.score,
+					T::SolutionImprovementThreshold::get(),
+				) {
+					Some(i + 1)
+				} else {
+					None
+				}
+			})
+			.or(Some(0))
```

Comment:
I think [`binary_search_by_key`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.binary_search_by_key) could be helpful here. From the docs:

> If the value is not found then `Result::Err` is returned, containing the index where a matching element could be inserted while maintaining sorted order.

Reply:
f5f4605